packages/react-devtools-shared/src/__tests__/store-test.js

```diff
@@ -1008,4 +1010,197 @@ describe('Store', () => {
       done();
     });
   });
+
+  describe('inline errors and warnings', () => {
+    test('during render are counted', () => {
+      function Example() {
+        console.error('test-only: render error');
+        console.warn('test-only: render warning');
+        return null;
+      }
+      const container = document.createElement('div');
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 1, warnings: 1}]]),
+      );
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example rerender={1} />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 2, warnings: 2}]]),
+      );
+    });
+
+    test('during layout get counted', () => {
+      function Example() {
+        React.useLayoutEffect(() => {
+          console.error('test-only: layout error');
+          console.warn('test-only: layout warning');
+        });
+        return null;
+      }
+      const container = document.createElement('div');
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 1, warnings: 1}]]),
+      );
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example rerender={1} />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 2, warnings: 2}]]),
+      );
+    });
+
+    test('during passive get counted', () => {
+      function Example() {
+        React.useEffect(() => {
+          console.error('test-only: passive error');
+          console.warn('test-only: passive warning');
+        });
+        return null;
+      }
+      const container = document.createElement('div');
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 1, warnings: 1}]]),
+      );
+
+      withErrorsOrWarningsIgnored(['test-only:'], () => {
+        act(() => ReactDOM.render(<Example rerender={1} />, container));
+      });
+
+      expect(store.errorsAndWarnings).toEqual(
+        new Map([[2, {errors: 2, warnings: 2}]]),
+      );
+    });
+
+    // TODO (inline errors) The fiber for the recorded error is considered to be not mounted by renderer.js#isFiberMounted.
```

Comment:
Thought a bit more about how we batch errors/warnings: Isn't it possible that we loose errors if backend/console flushes them but the fiber isn't mounted yet? Should we move the batching back into the renderer and hold onto errors/warnings on fibers that haven't mounted? Or maybe even just emit them and let the frontend decide what to do with that information?

Reply:
I'm working on moving the batching into the render now, like I mentioned at the end fo the day yesterday. Just didn't get it finished yet. Will keep you posted.

Reply:
From what it looks like this has been addressed i.e. we send warnings/errors for every fiber regardless of its mount state. Before we skipped them if the fiber was unmounted or wasn't mounted yet.

But I think the frontend should now check if a fiber is mounted when counting the errors/warnings? Would probably be a good to have a test that simulates a long running render where we flush the warnings explicitly before the Fibers of those errors get mounted.

Slightly related: Have we thought about what we want to do with errors of renders that are thrown away. For example, we render+error, suspend, render+error, commit. I guess right now we're logging both errors?

Though the errors are logged to the console so we probably should display them in devtools. Just thinking out loud here :smile:

Reply:
> But I think the frontend should now check if a fiber is mounted when counting the errors/warnings?

I also noticed that the frontend counts errors of filtered out components. That is slightly confusing since I wasn't able to navigate to hit with "Scroll to next error or warning".

Couple of solutions:
1. deactivate the filter if we attempt to scroll to an excluded Fiber
2. ignore warnings/errors from excluded Fibers

I think 2 was the initial plan for the MVP but I would prefer we never hide errors/warnings unless they're acknowledged by either clearing or filtering (not part of this PR).

Reply:
> 2. ignore warnings/errors from excluded Fibers

This seems like the right approach. We shouldn't override filters.